### PR TITLE
[JAVA][feature] add jackson support to retrofit2 library 

### PR DESCRIPTION
**Pull request by matthewmcgarvey**
### PR checklist

- [x] Read the [contribution guidelines](https://github.com/swagger-api/swagger-codegen/blob/master/CONTRIBUTING.md).
- [x] Ran the shell script under `./bin/` to update Petstore sample so that CIs can verify the change. (For instance, only need to run `./bin/{LANG}-petstore.sh` and `./bin/security/{LANG}-petstore.sh` if updating the {LANG} (e.g. php, ruby, python, etc) code generator or {LANG} client's mustache templates). Windows batch files can be found in `.\bin\windows\`.
- [x] Filed the PR against the correct branch: `3.0.0` branch for changes related to OpenAPI spec 3.0. Default: `master`.
- [] Copied the [technical committee](https://github.com/swagger-api/swagger-codegen/#swagger-codegen-technical-committee) to review the pull request if your PR is targeting a particular programming language.

### Description of the PR
Issue #7620 

Allows for the retrofit2 library to be used with jackson instead of gson.  Could have solved with templates but gson was hardcoded in the JavaClientCodegen as well as adding a template that depended on gson.

Sidenote: Has there been thought to reorganizing the library structure?  There are a lot of optional properties in JavaClientCodegen that have to do with specifically retrofit

### CODE REVIEW

File: modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache [hunk 1]
```diff
@@ -47,10 +45,15 @@ public class ApiClient {
...
+  {{#gson}}
+  private JSON json = new JSON();
+  {{/gson}}
+  {{#jackson}}
+  private ObjectMapper mapper = new ObjectMapper();
+  {{/jackson}}
 
   public ApiClient() {
-    apiAuthorizations = new LinkedHashMap<String, Interceptor>();
+    apiAuthorizations = new LinkedHashMap<>();
```
Comment (cbornet) [inline modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache]: The code should be compatible with JDK6 (for Android)

File: modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache [hunk 2]
```diff
@@ -154,40 +156,8 @@ public class ApiClient {
...
-    return this;
-  }
-
-  public ApiClient setLocalDateFormat(DateTimeFormatter dateFormat) {
-    this.json.setLocalDateFormat(dateFormat);
-    return this;
-  }
-
-  {{/jsr310}}
+{{#gson}}{{>gson/mappingConverterFactory}}{{/gson}}
```
Comment (cbornet) [inline modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache]: Can we put this into another class/file instead ?

Comment (cbornet) [reply modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache]: Eg : the JSON class

Comment (matthewmcgarvey) [reply modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache]: I have fixed the java6 issue, but I don't believe that this change will work.  The JSON class for the gson library was necessary for all the gson configuration but Jackson doesn't need any configuration in this circumstance, just to replace the gson logic in this class which is tied to the ApiClient

Comment (cbornet) [reply modules/swagger-codegen/src/main/resources/Java/libraries/retrofit2/ApiClient.mustache]: There is also mapper configuration for Jackson that should be moved to a JSON class.

**Comment by davidyscott** [comment]
It looks like matthewmcgarvey put a lot of work into this change, which I would also greatly appreciate seeing.  What's happened with this PR?

**Pull request closed without merging**